Add `min/2`, `max/2` and `min_max/2` to `Enum`.

myronmarston opened the pull request:
As discussed on the [mailing list](https://groups.google.com/forum/#!topic/elixir-lang-core/goQuIjs00YE).

The new 2nd argument is a function that is called when the enumerable is empty. It defaults to raising `Enum.EmptyError` to preserve existing behavior.

One thing I'm unsure of: should we add `min_by/3`, `max_by/3` and `min_max_by/3` as well?

josevalim commented on lib/elixir/lib/enum.ex:
```diff
@@ -1724,6 +1736,14 @@ defmodule Enum do
   end
 
+  defp reduce_handling_empty(enumerable, fun, empty_fallback) do
+    if Enum.empty?(enumerable) do
```
We cannot call `Enum.empty?` because it may force the enumerable to run twice. We need to use the same schema we used in `Enum.min_max`.

myronmarston replied:
Good catch.  Fixed.

josevalim commented:
> One thing I'm unsure of: should we add min_by/3, max_by/3 and min_max_by/3 as well?

@whatyouhide @lexmag - what do you think?

lexmag commented:
@josevalim I think they are very connected and it'll be confusing not to provide.

lexmag commented on lib/elixir/lib/enum.ex:
```diff
@@ -1315,17 +1315,21 @@ defmodule Enum do
-  def max(enumerable) do
-    reduce(enumerable, &Kernel.max(&1, &2))
+  @spec max(t, (() -> empty_result)) :: element | empty_result | no_return when empty_result: any
+  def max(enumerable, empty_fallback \\ fn -> raise Enum.EmptyError end) do
```
We should have a `when is_function(empty_fallback, 0)` guard here and in `min/2`.
Also it would be better to move the definition of default for `empty_fallback` to bodiless clause.

myronmarston replied:
Done and done, although I was hoping you could explain the rationale behind:

> Also it would be better to move the definition of default for `empty_fallback` to bodiless clause.

Why is that better?

lexmag replied:
> Why is that better?

Only for readability purpose: actual handling and guards separated from defaults.

whatyouhide commented:
I agree with @lexmag, it feels a bit confusing *not* to have them 😕

myronmarston commented:
Ok, I think I've addressed all the code review feedback but I still need to implement the `*_by` functions.  I ran out of time tonight but will try to get to it tomorrow.

whatyouhide commented on lib/elixir/lib/enum.ex:
```diff
@@ -1419,17 +1424,22 @@ defmodule Enum do
-    reduce(enumerable, &Kernel.min(&1, &2))
+  @spec min(t, (() -> empty_result)) :: element | empty_result | no_return when empty_result: any
+  def min(enumerable, empty_fallback \\ fn -> raise Enum.EmptyError end)
+  def min(enumerable, empty_fallback) when is_function(empty_fallback, 0) do
```
This is a bit nitpicky, sorry, but we usually leave an empty line between a bodyless clause and its implementation 😄

myronmarston replied:
Done.

whatyouhide commented on lib/elixir/lib/enum.ex:
```diff
@@ -1485,28 +1495,24 @@ defmodule Enum do
   that was found is returned.
 
-  Raises `Enum.EmptyError` if `enumerable` is empty.
+  Calls the provided `empty_fallback` and returns its value if `enumerable`
```
"the provideded `empty_fallback` **function**", wdyt?

myronmarston replied:
Done.

myronmarston commented:
OK, I've addressed all the feedback so far and added `min_by/3`, `max_by/3` and `min_max_by/3`.  I wasn't 100% sure on this but it looked like the clauses of `min_by/3` and `max_by/3` specifically for lists were superfluous (perhaps vestigal from a prior implementation?) so I went ahead and removed them but did so in a separate commit that I can easily remove from this PR if there's a reason to keep those clauses (which, if there is, I'd like to understand!).

This PR is now 7 commits.  Let me know if you want me to squash it into fewer to prep for merge.  It probably still makes sense to keep 882eb0ced63d05e5c72910e3c1e531ab3a91afea as its own commit but the rest of this could probably be one commit if there's a preference for that.

josevalim commented on lib/elixir/test/elixir/enum_test.exs:
```diff
@@ -314,16 +331,44 @@ defmodule EnumTest do
+    assert_runs_enumeration_only_once(&Enum.min(&1, fn -> nil end))
+  end
+
+  defp assert_runs_enumeration_only_once(enum_fun) do
```
Let's move this to either the top or the bottom of the file. :)

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

josevalim commented:
@myronmarston Thank you! the reason we inlined lists was for performance. I have pushed a commit that bring those improvements back but at the level of the reduce_handling_empty function. :) I have also split the reduce_handling_empty in two, so we don't need to use references in min_max_by.

myronmarston commented:
> Thank you! the reason we inlined lists was for performance. I have pushed a commit that bring those improvements back but at the level of the reduce_handling_empty function. :) I have also split the reduce_handling_empty in two, so we don't need to use references in min_max_by.

:+1: Your changes look great :).  I can understand how the new implementation you pushed that delegates to `:lists.foldl/3` would be faster, but I'm a bit confused about how the old implementations could be faster.  Here's the code for `min_by/2` as it existed before this PR:

``` elixir
  def min_by([h | t], fun) when is_function(fun, 1) do
    reduce(t, {h, fun.(h)}, fn(entry, {_, fun_min} = old) ->
      fun_entry = fun.(entry)
      if(fun_entry < fun_min, do: {entry, fun_entry}, else: old)
    end) |> elem(0)
  end

  def min_by([], _fun) do
    raise Enum.EmptyError
  end

  def min_by(enumerable, fun) when is_function(fun, 1) do
    result =
      reduce(enumerable, :first, fn
        entry, {_, fun_min} = old ->
          fun_entry = fun.(entry)
          if(fun_entry < fun_min, do: {entry, fun_entry}, else: old)
        entry, :first ->
          {entry, fun.(entry)}
      end)

    case result do
      :first       -> raise Enum.EmptyError
      {entry, _} -> entry
    end
  end
```

The implementations for the non-empty list and other enumerables is nearly identical -- both delegate to `reduce/3` with a nearly identical function.  The only differences I can see are:

* The list clause peels off the first element and applies the function to get the first accumulator value instead of using `:first`
* The reduction has one less iteration
* The passed function has only clause, since there's no need for the `entry, :first` clause
* The result is handled by a `case` statement instead of `elem(0)`

From my current understanding of Elixir, it doesn't seem like the perf should be noticeably different (and in my own code I'd opt to reduce the duplication as a result), but I'm clearly missing something.  In the interest of improving my intuition about Elixir performance, can you tell me which of the differences makes for a noticeable perf difference?

josevalim commented:
I agree with you. The implementation for min_by likely did not provide relevant gains. Maybe it did at some point, but not the latest version you removed.
